### ISSUE TITLE
[Hunter] Shot queueing

### ISSUE THREAD

**Issue by carolusr3x** [Atlantiss/NetherwingBugtracker#2339]
**Description**: 
Shots with a cast time (Steady, Aimed, Multi) should not interrupt auto shot cast and instead be queued to start after auto shot, as explained [here](https://github.com/Schaka/TBC-research/issues/28) (links from EJ inside for proof).

**Current behaviour**:
Cast shots interrupt auto shot.

**Expected behaviour**: 
Shots used during auto shot cast should trigger the GCD but not start casting until auto shot has fired.

**Comment by mhdemn** [comment]
What is explained in the post isn't the same as what you're saying tho, is it?
You're saying shots with a cast time should automatically recognize you're firing an auto shot and wait for it to finish to start casting. Then why hunters back in retail TBC came up with the 1:1 rotation macro? Maybe it was because WoW's engine doesn't work like that by default and the macro did that for you? TBC engine was designed so when you pressed a casted shot such as Steady Shot you started casting it, without any regard of how soon you were gonna fire another Auto Shot.

**Comment by carolusr3x** [comment]
The macro doesn't work either, you can't do the 3:2 rotation properly without this because you clip auto shot. The 3:2 rotation lets slower weapons compare in dps to fast weapons with 1:1 rotation.

Maybe my explanation isn't correct but you can compare to other private servers such as warmane, where shot queueing is implemented correctly (afaik).

**Comment by mhdemn** [comment]
Yes, I was typing about that right now. 1:1 works fine but 3:2 just spams Steady Shot without ever letting 1 out of 2 autos out. 
However, if this was fixed in 2.4, this is a core change and perhaps it's something that can't be just implemented from the developers side. I mean, I think 2.4.3 client is designed so 3:2 rotation macro doesn't work, and making this macro work could require something bigger than a server fix + restart.

**Comment by carolusr3x** [comment]
According to the EJ thread in the link this was implemented in 2.3 iirc. This works on warmane-outland with the same 2.4.3 client.

**Comment by mhdemn** [comment]
I see. However, rn I'm a little confused cause I decided to give record 3:2 function in NW currently and somehow it works sometimes. So I dunno what the hell's up here

**Comment by mhdemn** [comment]
Okay, as seen on this [video](https://youtu.be/ycklqA9DEKE) in server revision 2255 autoshot gets clipped sometimes, sometimes it doesn't while all I do is smash the macro all the time. Couldn't reproduce the moment where I got it to work steadily, but this sums it up.

My point is this is something that IMO should be in TBC through all content, and not only until 2.4. However, if how it functions currently in NW isn't how it's supposed to (either macro should work and do 3:2 rotations or never work and clip Autos always), it could use some fixing.

For clarification, [this](https://youtu.be/BWGL0jbMF1w) is how 3:2 rotation looks like:
[AA-SH-SH-AA-SH]-AA-SH-SH-AA-SH-AA-SH-SH-AA-SH-AA-SH-SH-AA-SH

AA=Auto Shot
SH=Steady Shot

**Comment by carolusr3x** [comment]
Actually afaik Aimed Shot should reset the Auto Shot timer (like Slam resets the swing timer for warriors), Auto shouldn't fire immediately after Aimed Shot. Not sure if it's related but that's also a bug.

**Comment by mhdemn** [comment]
did you notice 1:1 macro not working properly sometimes since not long ago @carolusr3x ? or is it my lag and my new weapon's speed

**Comment by mhdemn** [comment]
Furthermore, shots that are casted thru a macro like

#showtooltip
/cast [mod:ctrl] !Arcane Shot(Rank 1); [mod:shift][nomod] !Arcane Shot
/stopattack [mod:ctrl]
/cleartarget [mod:ctrl]
/targetlasttarget [mod:ctrl]
/petpassive [mod:ctrl]
/startattack [nomod]
/petdefensive [nomod:ctrl]
/petattack [nomod]

or even something simple like

#showtooltip
/cast Serpent Sting
/startattack 
/petdefensive

will get stuck sometimes like an active Heroic Strike waiting to empower next autoattack; and until your next auto shot goes off they won't fire but stay highlighted.

**Comment by mhdemn** [comment]
2304 (same as above)

TL:DR: another bug that happens when messing with traps/target switching/actively shooting that will stop the hunter from autoattacking until combat is over. Feign Death combat-exiting doesn't work, the mobs that caused the bug need to die, then the bug is over. Kind of a big deal if this happened during a raid boss.

Since #2452 seems to be the pet bug thread now, this may as well be the pet owners' one.

Wanted to add this one since it's really annoying but it can escalate to a game breaking bug; but I'm not seeing any open issue about it.

In this [video](https://youtu.be/iPCVR2kdpeo) I thought of hitting record when I already had the bug active.

This bug will prevent the hunter from autoattacking it's target in any way, and clearing target, feigning death to exit combat or spamming the /startattack macro won't fix it. From the times this has happened, even though I couldn't record all these trials in video, I learned that only when the combat (during which the bug occured) ends, the hunter recovers the capacity to auto shoot.

This happened to me before very rarely, but today after having to trap targets in dungeons pull after pull, and messing with freezing traps and distracting shots all the time it seems the trigger for this bug is somewhat related to laying down traps in some specific situation. I'll leave record going next time I gather people for a dungeon and hopefully catch something that can clarify it, because I can't identify the cause.

Again, couldn't get a video of everything I wanted to record, but I'm gonna report the issue anyway, just in case you're aware or some other player knows more about this.

**Comment by carolusr3x** [comment]
can't comment on the 1:1 macro since I'm doing my rotations manually but I noticed my instant shots sometimes get stuck like that even without a macro, they won't go off until I stand still and wait for auto shot. also I can get 3:2 to work sometimes but if I spam it too hard I just keep casting steady shot without any auto shots going off.

**Comment by mhdemn** [comment]
Yea, meant to add that my multishot is also getting stuck now, and its not in a macro.

Easy to reproduce this when you try to cast a shot on a LoS'd target (the highlighted and blocked shot), not sure if its the exact same thing though.

I relate to that 3:2 macro behaviour too.

**Comment by amarathTESTER** [comment]
Take at my testing on nightbane. For this, they did something that caused a steady spam (typical thats an gcd formula issue or an auto swing issue).

https://github.com/SunwellWoW/Sunwell-TBC-Bugtracker/issues/590

Is the gcd 1.5 ? Only effecting spells for haste? 
Does the autoswing timer always start when asked and not on in gcd and goes off as long as when the swing occurs you are not casting (hit within the same server update [not spell queue] as a kc)?

Best of luck.

**Comment by ta55el** [comment]
https://www.ownedcore.com/forums/world-of-warcraft/world-of-warcraft-exploits/114539-hunter-auto-shot-bug-dps-increase.html

**Comment by letterix** [comment]
@aresbrutus What research is needed here? 

3:2 and 1:1 macros and clipping were a thing throughout the TBC raiding scene and it affected the tuning blizzard did. If those are not in place, the class will function and scale entirely different. There is not a lack of sources on this. Whether this is solved via macros or made core, the functionality needs to be there.

**Comment by Zarant** [comment]
I dont know what kind of research we still need to validate this report, I'm gonna paste the same report I did a month or so ago. The issue was fixed at some point but right now it's broken again.

During patch 2.3, spell queuing mechanics were added to the game, this change affected most spells in the game, including auto shot.

> Client spell cast requests are now sent to the server even if your player is already casting another spell. This eliminates the need for /stopcasting in macros to compensate for latency. 

Source: http://wowwiki.wikia.com/wiki/Patch_2.3.0

Once the 0.5s Auto Shot cast time starts, it shouldn't be interrupted by other shots, spell casts performed during the Auto Shot wind up time should be queued and handled by the server immediately after the auto shot goes off. Before patch 2.3, specially using fast attack speed weapons or under the effect temporary haste buffs hunters had to compensate for latency by casting steady shot a split second before the auto shot went off. Patch 2.3 added a huge QoL change that completely kicked the latency compensation component out of the DPS rotation.

>In short:
1 - GCD is client only, and starts immediately upon sending the cast to the server
2 - Auto Shot can't be interrupted once the cast starts by other casts
3 - The next cast is "queued up" and began immediatetly after Auto Shot
4 - The server doesn't have any cooldown on Steady Shot. It counts on the client handling that through the GCD

Source: https://web.archive.org/web/20080211131225/http://elitistjerks.com:80/f31/t11420-hunter_beast_mastery_bible/p109/ - see the last post, from January 28th, 2008 (patch 2.3.3)

**Comment by letterix** [comment]
@Zarant Did you see the 3:2 rotation work consistently before on the beta?

**Comment by Zarant** [comment]
I tested during phase 1 and 2, and it was working properly with multi-shot, I couldn't test steady shot since I had no access to level 62 characters.

**Comment by Nonamesleftffs** [comment]
https://www.youtube.com/watch?v=5oPveXRCXp8&t=2s

I made this video a few weeks ago to show hunter shot selections.. The 2:3 shot rotation is the 3rd one shown.

Currently it doesn't work on Netherwing, but it should. You intentionally delay auto shots occasionally to fit in a second steady shot between 2 autop shots, but the way it SHOULD work, is that the auto shot should go off immediately after the second steady shot (immediately after the delay).. Here, on Netherwing, the 2nd auto shot just doesn't happen, and you end up spamming steady shot with absolutely no auto shots at all.. Kinda like playing a mage, and casting nothing but fireball.

**Comment by DragunovUA** [comment]
any news?

**Comment by Wolffenstein** [comment]
Retest in rev 2527.

**Comment by Zarant** [comment]
Just tested it, it kinda works, there's only one small thing missing, the 3:2 rotation doesn't work with this current implementation. 
If you're trying two steadies back to back, auto shots should take precedence, as long as it's off cooldown, right now you can keep queuing steady shots back to back indefinitely.

**Comment by Wolffenstein** [comment]
Please recheck in revision 2537.

**Comment by Zarant** [comment]
Feels better, but it's still unreliable, I tested with a 2.0 speed weapon which leads to a 1.45s attack speed, if you consider the fact that steady is on a 1.224 cast time and the auto shot wind up time is 0.5s, auto shot should always be available after a steady shot and I still get 2 back to back steady shot casts every now and then.

I recorded the test I did:
https://youtu.be/_Z1vkE-FguE

**Comment by Zarant** [comment]
I also found out that the whole shot queuing thing sometimes falls apart with aimed shot since aimed shot is supposed to trigger an auto shot CD. 
If you queue a a steady after an aimed shot, your steady shot will be delayed long enough and will only be cast after the next auto, instead of queuing directly after the aimed shot.

**Comment by Zarant** [comment]
Instant shots are also being handled in the same way, including scatter shot and silencing shot. #3738

**Comment by Zarant** [comment]
I'll try to break down the auto shot mechanics and all its nuances as much as possible:

1. Auto shot is a spell with a 0.5s cast time and its cooldown is the hunter's ranged attack speed minus 0.5
2. Instant cast spells don't interact with auto shot in any way.
3. Auto shot can only be cast if you're not currently casting any spell and only while standing still.
4. Casting any spell with a cast time should interrupt the 0.5s auto shot cast time.
5. Steady shot and multi-shot are the only exceptions to rule number 4, both spells don't interrupt auto shot.
6. If you cast steady shot or multi-shot while an auto shot is mid cast, your next steady shot or multi-shot should be queued and fired as soon as the 0.5s auto shot cast time finishes.
7. Whenever you finish casting an aimed shot, your auto shot cooldown is reset.

Now this next part is crucial to understand why the 3:2 rotation works the way it does:

8. If you queue a steady shot or multi-shot during the auto shot wind up time, your GCD starts when the game client send the spell cast to the server.
9. Steady shot has a cast time shorter than a GCD, since ranged haste doesn't affect GCD.
10. Spell queuing should respect GCD limitations, meaning that you can't cast 2 spells back to back if both of them have a cast time shorter than a GCD.
11. If you cast a steady shot during the auto shot wind up time, the spell triggers a GCD before the spell cast begins, allowing you to cast 2 back to back steady shots.
12. You can't cast 2 back to back steady shots if auto shot is off cooldown by the time the first steady shot cast finishes. Auto shot should always take precedence over steady/multi shot.
13. You can't queue a multishot after a steady shot if auto shot is off cooldown by the time the first steady shot cast finishes.

**Comment by DragunovUA** [comment]
bump

**Comment by DragunovUA** [comment]
up

**Comment by Zarant** [comment]
I was watching some old tbc videos and found out that the auto shot > multi-shot queue behavior got changed somewhere in between patch 2.3 and patch 2.4.3

https://youtu.be/LrZ_qgp917c?t=315 Hunter have auto shot turned on, stops for a multi-shot, if you advance it frame by frame you can see auto shot happening after multi-shot, however just judging by the gear of the opponents he faced, this video was recorded in season 2 (patch 2.1)


This video, from S3 (patch 2.3) shows the multi-shot queuing:
https://youtu.be/m_cVI65BHn0?t=1067 - Same situation, hunter has auto shot turned on, stops to fire a multi-shot, auto shot go off first, multi shot fires 0.5s later, you can see the multi-shot being delayed significantly here.


https://youtu.be/wCBmCVOBhQY?t=269 - clip from S4, multi shot goes first, auto shot 0.5 later
https://youtu.be/wCBmCVOBhQY?t=287 - same video, same thing happening at a different time


https://youtu.be/h0eVrsYItpI?t=301 - auto shot is off cooldown and auto shot is turned on, the hunter stops to cast a multi-shot, instead of waiting for a auto shot to go off, multi shot is fired right away and if you pay close attention on the combat log, auto shot is fired 14 frames later (~0.467 seconds), by his gear you can see that this was recorded in S3, maybe S4 (after patch 2.3)

As you can see, from the S4 video, the auto shot queuing should be limited only to steady shot after patch 2.4.

**Comment by Zarant** [comment]
Since the last update (revision 2816) shot queuing is completely broken and it's not working at all.

**Comment by Godmatik** [comment]
Fixed a long time ago
